3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/VeroAdapter.scala

```diff
@@ -0,0 +1,167 @@
+
+// json4s
+import org.json4s._
+import org.json4s.jackson.JsonMethods._
+
+// Iglu
+import com.snowplowanalytics.iglu.client.{Resolver, SchemaKey}
+
+// Joda Time
+import org.joda.time.{DateTime, DateTimeZone}
+import org.joda.time.format.{DateTimeFormat, DateTimeFormatter}
+
+// This project
+import com.snowplowanalytics.snowplow.enrich.common.loaders.CollectorPayload
+//import com.snowplowanalytics.snowplow.enrich.common.utils.{JsonUtils => JU}
+
+import scala.util.{Failure, Success, Try}
+
+/**
+ * Transforms a collector payload which conforms to
+ * a known version of the Vero webhook
+ * into raw events.
+ */
+object VeroAdapter extends Adapter {
+
+  // Vendor name for Failure Message
+  private val VendorName = "Vero"
+
+  // Expected content type for a request body
+  private val ContentType = "application/json"
+
+  // Tracker version for an Vero webhook
+  private val TrackerVersion = "com.getvero-v1"
+
+  // Schemas for reverse-engineering a Snowplow unstructured event
+  private val EventSchemaMap = Map(
+    "bounced"      -> SchemaKey("com.getvero", "bounced", "jsonschema", "1-0-0").toSchemaUri,
+    "clicked"      -> SchemaKey("com.getvero", "clicked", "jsonschema", "1-0-0").toSchemaUri,
+    "delivered"    -> SchemaKey("com.getvero", "delivered", "jsonschema", "1-0-0").toSchemaUri,
+    "opened"       -> SchemaKey("com.getvero", "opened", "jsonschema", "1-0-0").toSchemaUri,
+    "sent"         -> SchemaKey("com.getvero", "sent", "jsonschema", "1-0-0").toSchemaUri,
+    "unsubscribed" -> SchemaKey("com.getvero", "unsubscribed", "jsonschema", "1-0-0").toSchemaUri,
+    "user_created" -> SchemaKey("com.getvero", "created", "jsonschema", "1-0-0").toSchemaUri,
+    "user_updated" -> SchemaKey("com.getvero", "updated", "jsonschema", "1-0-0").toSchemaUri
+  )
+
+  /**
+   * Converts a payload into a single validated event
+   * Expects a valid json returns failure if one is not present
+   * For "user_updated" events the field is named "action" instead of "type"
+   * in which case the event type will be set to Some("user_updated")
+   *
+   * @param json Payload body that is sent by Vero
+   * @param payload The details of the payload
+   * @return a Validation boxing either a NEL of RawEvents on
+   *         Success, or a NEL of Failure Strings
+   */
+  private def payloadBodyToEvent(json: String, payload: CollectorPayload): Validated[RawEvent] = {
+
+    val parsed = Try(parse(json))
+    parsed match {
+      case Success(parsed) => parsed
+      case Failure(e)      => return s"$VendorName event failed to parse into JSON: [${e.getMessage}]".failureNel
+    }
+    val eventType = (parsed.get \ "type").extractOpt[String].orElse(Some("user_updated"))
+    val formattedEvent = eventType match {
+      case Some("user_updated") => parsed.get
+      case _ => {
+        cleanupJsonEventValues(
+          parsed.get,
+          ("type", eventType.get).some,
+          s"${eventType.get}_at"
+        )
+      }
+    }
```

Comment:
you should be able to do:

```scala
val formattedEvent = parsed.map { p =>
  if (eventType == "user_updated") p
  else cleanupJsonEventValues(p, ("type", eventType).some, s"$eventType_at")
}
```

Reply:
In general, it's good practice to avoid `.get`

Reply:
Similarly here, 'formattedEvent' will return wrapped in Try() but needs to be a JValue to be passed into the function 'reformatParameters' and then into 'toUnstructEventParams'.

Reply:
you can do a for comprehension:

```scala
val parsed = Try(parse(json)) match {
  case Success(p) => p.successNel
  case Failure(e) => s"$VendorName event failed to parse into JSON: [${e.getMessage}]".failureNel
}
val eventType = parsed.toOption.flatMap(p => (p \ "type").extractOpt[String]).getOrElse("user_updated")
for {
  parsedEvent <- parsed
  formattedEvent =
    if (eventType == "user_updated") parsedEvent
    else cleanupJsonEventValues(parsedEvent, ("type", eventType).some, s"$eventType_at")
  reformattedEvent = reformatParameters(formattedEvent)
  params = toUnstructEventParams(TrackerVersion, toMap(payload.queryString), schema, reformattedEvent, "srv")
  schema <- lookupSchema(eventType.some, VendorName, EventSchemaMap).toValidationNel
  rawEvent = RawEvent(api = payload.api, parameters = params, contentType = payload.contentType, source = payload.source, context = payload.context)
} yield rawEvent
```